src/core/moves/MetropolisHastingsMove.cpp

```diff
@@ -259,10 +260,90 @@ void MetropolisHastingsMove::performMcmcMove( double prHeat, double lHeat, doubl
     const RbOrderedSet<DagNode*> &affected_nodes = getAffectedNodes();
     const std::vector<DagNode*> nodes = getDagNodes();
     
+    
+    
+    // --------------------------
+    //
+    //     DEBUG (BEGIN)
+    //
+    // --------------------------
+#ifdef DEBUG_MCMC
+    double ln_posterior_before_move = 0.0;
+    for (size_t i = 0; i < nodes.size(); ++i)
+    {
+        // get the pointer to the current node
+        DagNode* the_node = nodes[i];
+        ln_posterior_before_move += the_node->getLnProbability();
+    }
+    for (RbOrderedSet<DagNode*>::const_iterator it = affected_nodes.begin(); it != affected_nodes.end(); ++it)
+    {
+        DagNode *the_node = *it;
+        ln_posterior_before_move += the_node->getLnProbability();
+    }
+    for (size_t i = 0; i < nodes.size(); ++i)
+    {
+        // get the pointer to the current node
+        DagNode* the_node = nodes[i];
+        the_node->touch();
+    }
+    for (RbOrderedSet<DagNode*>::const_iterator it = affected_nodes.begin(); it != affected_nodes.end(); ++it)
+    {
+        DagNode *the_node = *it;
+        the_node->touch();
+    }
+    double ln_posterior_before_move_after_touch = 0.0;
+    for (size_t i = 0; i < nodes.size(); ++i)
+    {
+        // get the pointer to the current node
+        DagNode* the_node = nodes[i];
+        ln_posterior_before_move_after_touch += the_node->getLnProbability();
+    }
+    for (RbOrderedSet<DagNode*>::const_iterator it = affected_nodes.begin(); it != affected_nodes.end(); ++it)
+    {
+        DagNode *the_node = *it;
+        ln_posterior_before_move_after_touch += the_node->getLnProbability();
+    }
+    for (size_t i = 0; i < nodes.size(); ++i)
+    {
+        // get the pointer to the current node
+        DagNode* the_node = nodes[i];
+        the_node->keep();
+    }
+    for (RbOrderedSet<DagNode*>::const_iterator it = affected_nodes.begin(); it != affected_nodes.end(); ++it)
+    {
+        DagNode *the_node = *it;
+        the_node->keep();
+    }
+    if ( fabs(ln_posterior_before_move - ln_posterior_before_move_after_touch) > 1E-6 )
+    {
+        throw RbException("Issue in '" + proposal->getProposalName() + "' on '" + nodes[0]->getName() + "' before move because posterior of " + ln_posterior_before_move + " and " + ln_posterior_before_move_after_touch + ".");
+    }
+#endif
+    // --------------------------
+    //
+    //     DEBUG (END)
+    //
+    // --------------------------
+    
+    
+    
+    
+    
+    
     // Propose a new value
     proposal->prepareProposal();
-    double ln_hastings_ratio = proposal->doProposal();
-    
+    double ln_hastings_ratio = RbConstants::Double::neginf;
+    try {
+        ln_hastings_ratio = proposal->doProposal();
+    }
+    catch (RbException e)
```

Comment:
I don't think catching exceptions by value is safe.  If you do 
``` c++
catch (const RbException& e)
```
I think that is safe.

Reply:
Hm, I'm not aware of this being an issue. If so, then should we check the entire code base?

Reply:
This is my example of catching exceptions by reference, instead of by value.

Reply:
OK, I changed these in commit a4f5cf43.